### PR TITLE
Generative model of the OASIS brain data set using StyleGAN2

### PR DESCRIPTION
**Pull request by yourcousinbob**
Hi Shakes and any COMP3710 tutors,

This is my pull request for a Tensorflow implementation of StyleGAN2 and other utility functions with full documentation designed to be used on the OASIS brain dataset. Unfortunately I was unable to produce consistently decent images within the given time frame however I have outlined a summary of my attempts to do so. There is a total of 5 python files. 

* `driver.py` contains the main function to begin training as well as argument parsing options for easy loading and clearing of previous training runs. 
* `generator.py` and `discriminator.py` hold the classes representing the respective models with easy tuning of parameters at the top of each file. 
* `gan.py` holds the class representing the StyleGAN2 network and any utility functions and parameters associated with it. 
* `layers.py` which holds any modified Tensorflow layers that were used in training. 

 The images folder contains images used in the README. 

I would like to add that it seems like something might be incompatible with this code and the rangpur cluster as I could not get it to produce anything but black images with a nan loss despite the code and environment being the same as my local machine. I attribute this in large part why I ran out of time to train the models.

If there is any issues or fixes you find in reviewing this code let me know.

Thanks,
Bobby Melhem 45829211

### CODE REVIEW

**Review by WeiDaiwow — COMMENTED** [review]
Have both test and driver file
Good commits 
Nice readme file
Good design and nice implementation
Nice comments
Nice description and explanation of the algorithm
Correct pull request with the description
nice visualization and plots
Dependencies included
Nice Markdown
Nice reference

Although the quality of the generated outputs is not very high, they do show decent patterns and variations. The final results are acceptable to me.

Please show more plots during the training under different epochs.

**Comment by shakes76** [comment]
## TF/Torch Usage
Adequate use and implementation
## Good Practice (Design/Commenting)
Good spacing and comments
Header blocks
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
